modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js

```diff
@@ -230,6 +232,41 @@ const FrontendDataSet = ({
 		setItems(dataSetData.items);
 	}
 
+	useEffect(() => {
+		if (!initialFilters?.length) {
+			return;
+		}
+
+		let dispatch = false;
+
+		const filterPromises = initialFilters.map((filter) => {
+			if (!filter.cxFilterURL) {
+				return Promise.resolve(filter);
+			}
+
+			dispatch = true;
+
+			return import(
+
+				/* webpackIgnore: true */
+				filter.cxFilterURL
+			).then((module) => {
+				filter.cxFilterImplementation = module['default'];
+
+				return filter;
+			});
+		});
```

Comment:
The idea behind [getRenderer](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/renderer/getRenderer.ts) was to avoid having this logic duplicated in every CX use. Instead we would do something like
```js
const renderer = getRenderer({
   type: 'clientExtension',
   url: cxFilterURL,
 });
```
and from then on use `renderer.htmlElementBuilder`.

Of course, that's not directly applicable in this case, as we need the whole `cxFilterImplementation`.

Would it maybe make sense to extend `getRenderer` and support something like
```js
const renderer = getRenderer({
   type: 'fdsFilterClientExtension',
   url: cxFilterURL,
 });
```
and have it return the full implementation?

Reply:
Yeah, I saw the getRenderer function but since this wasn't a renderer I thought it didn't make sense. But I agree we should not have one useEffect per CX type.

Another difference I see here is that the flow for handling the CX is:

1. Start rendering the FDS
2. An async request is made to load the CXs
3. The render continues without the CXs and uses spinners or empty strings where the CX would be used
4. When the CXs arrive, I dispatch a FILTERS_CHANGED event and a re-render is triggered, removing all spinners and empty string for free.

In the case of renderers, this flow happens per cell so it's quite different.

Maybe we could make a mix of the two techniques: send asyn requests for all external stuff at the very beginning, then firing re-renders when things arrive. That way we wouldn't need a `getRenderer` function, but would instead pass an optional prop to cell with the renderer set to something or set to null...

Would something like that work?

Reply:
It should the same flow for cells and for views, they are both initially rendered as spinners and resolved as React components in the end. There is a complication with reducer for views, but that's effectively just to bundle up and assure order of state changes.